feat: Generate link previews in StatusChatInput

alexjba opened the pull request:
### What does the PR do

Closing #12181 

Link previews are generated inside StatusChatInput based on the input text. The link preview area is above the text and this area will accommodate image previews and link preview. It's a scrollable area with a fade out effect when scrolling the cards out of the view. Once a link is detected in the chat input area a new preview card will be visible, the link will be styled using blue and underlinings. Once the link preview is hovered the corresponding text hyperlink is highlighted.

Changes:
1. Adding the newly designed components
  - LinkPreviewMiniCard - This component is similar to LinkPreviewCard, but has less info and a different format. There is also some additional functionality attached to this card: close button, reload button and hovered state. The LinkPreviewMiniCard can have multiple states, based on content type and loading state.

 - ChatInputLinksPreviewArea - This component is a wrapper for LinkPreviewMiniCard and StatusChatInputImageArea. The purpose of this component is to arrange the cards in a row layout and provide scrolling behaviour. This component also has an opacity mask that will provide a fade out appearance when the items are scrolled out of view.

 - TextEditHyperlinksFormatter - This component is attached to a text edit and will provide the hyperlinks formatting based on the urls detected in the text. The URL detection is implemented in the backend and this component will receive the urls as a model. The formatting is done using regular expressions and the hyperlinks are inserted automatically in the TextEdit component. The purpose of this component is to insert the hyperlinks, remove the hyperlinks and update the hyperlinks style when a specific hyperlink needs to be highlighted.

 2. Updating the link previews model. The model is updated to support insert/update and remove and the resetModel approach is removed. The resetModel is removed because of UI flickering when multiple links are inserted. Another update in the model is related to `remove card` functionality. The implementation is done using a newly introduced `immutable` role name. This property specifies if the row can be updated or not. For the remove card feature we need to remove the link and make sure it's not added back to the model once the link preview detection is triggered again for the same text. Once the link is removed from input area and added again, the link preview will be visible once again.

3. Adding storybook pages for the new components

4. Update ShapeRectangle to support all corner radius configuration and replace the rectangle opacity mask used to draw the callout card with ShapeRectangle. This was needed to support callout cards with dashed border.

5. Other small UI fixes (paddings, arrangements)

### Affected areas

StatusChatInput (URLs, emoji, mentions, images)
Sending images and links
MintedTokenView ( Empty placeholder when no assets; dashed rounded rectangle )
ProfileShowcase ( drop area background - dashed rounded rectangle )
ProfileSocialLinks ( empty placeholder when no links; dashed rounded rectangle )
Wallet history view (no transactions card - where "Activity for this account will appear here" is presented )
### Screenshot of functionality (including design for comparison)

- [x] I've checked the design and this PR matches it

#### Typing URLs:

https://github.com/status-im/status-desktop/assets/47811206/443c92a8-fc84-4011-943f-b9ac53f5f035

#### Handling multiple URLs - scroll + highlight + changing the URL + removing urls

https://github.com/status-im/status-desktop/assets/47811206/815c5e41-8e46-4f34-90cc-3446175f4e13

#### Handling images + links

https://github.com/status-im/status-desktop/assets/47811206/19f4c00b-4f05-4eb9-82cb-d5bbcb4f164b

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | b40a5ad8 | [#1](https://ci.status.im/job/status-desktop/job/branches/job/linux/job/x86_64/job/tests-statusq/job/PR-12262/1/) | 2023-09-26 11:36:11 | ~6 min | `tests/statusq` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/branches/job/linux/job/x86_64/job/tests-statusq/job/PR-12262/1//consoleText) |

caybro reviewed (commented):
Awesome work! Just some minor things inline

alexjba commented on ui/imports/shared/controls/chat/LinkPreviewMiniCard.qml:
```diff
@@ -0,0 +1,239 @@
+    implicitHeight: 64
+    verticalPadding: 15
+    horizontalPadding: 12
+    visible: true
```
Leftovers. Will remove

alexjba commented on ui/imports/shared/controls/chat/LinkPreviewMiniCard.qml:
```diff
@@ -0,0 +1,239 @@
+    required property url favIconUrl
+    required property url thumbnailImageUrl
+    required property int previewState
+    required property int type
```
I guess you're right. The required properties depend on the card type. In this case I can mark only the common properties as required

alexjba commented on ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml:
```diff
@@ -0,0 +1,153 @@
+        }
+    }
+
+    LinearGradient {
```
I'll give it a try. It should work. Thanks! 👍

alexjba replied:
Works great! Thanks

alexjba commented on ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml:
```diff
@@ -0,0 +1,153 @@
+                        required property int thumbnailWidth
+                        required property int thumbnailHeight
+                        required property string thumbnailUrl
+                        required property string thumbnailDataUri
```
That's right. I'll clean this up

alexjba commented on src/app/modules/shared_models/link_preview_model.nim:
```diff
@@ -105,33 +109,106 @@ QtObject:
+    for it in self:
+      if(it == url):
+        return true
+    return false
```
That will simplify things 👍

igor-sirotin reviewed (approved):
Amazing! Thank you! 🎉 

Just in case, did you test message editing? We use `StatusChatInput` there too. 
Editing link previews is not possible yet, but would be good to make sure that nothing broke here with these changes.

caybro commented on src/app/modules/shared_models/link_preview_model.nim:
```diff
@@ -105,33 +109,106 @@ QtObject:
+    let parentModelIndex = newQModelIndex()
+    defer: parentModelIndex.delete
+    self.beginInsertRows(parentModelIndex, self.items.len, self.items.len + itemsToInsert.len - 1)
+    self.items = concat(self.items, itemsToInsert)
```
I'd say... we should? So that the order of URLs in the text matches the order of previews?

alexjba replied:
Done

igor-sirotin commented on ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml:
```diff
@@ -0,0 +1,153 @@
+                        communityName: ""               //TODO: add community info when available
+                        channelName: ""                 //TODO: add community info when available
+
+                        thumbnailImageUrl: thumbnailDataUri.length > 0 ? thumbnailDataUri : thumbnailUrl
```
I think it should be the opposite.

```suggestion
                        thumbnailImageUrl: thumbnailDataUri.length == 0 ? thumbnailDataUri : thumbnailUrl
```

We should always prefer `thumbnailUrl` over `thumbnailDataUri`. The URL is pointing to a local media server.
- The former will be present when the preview is part of a Message. `daraUri` will never be appear in this case.
- The latter is used when the preview is shown in MessageComposer, because there's nowhere to point the URL to. Though I think we might have a URL in this case sometimes as well.

Also, I didn't look into the message previews part, please check that it's fine there too.

noeliaSD reviewed (approved):
Amazing job @alexjba !

Just added minor minor comments / questions.

alexjba commented on ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml:
> * I see the background area in white just in `storybook`. It looks ok in the app, but can you double check there is everything right?
It's a storybook issue. We need to use `Style.changeTheme` to change the theme. I'll fix it
> * I see that if you paste the same link more than once, the minicard is shown just only 1 time. Is that correct?
Yes

micieslak reviewed (changes requested):
First of all - a lot of good job!

Anyways, I've found several issues:

1. When editing message and deleting link, it ends up with segfault:

[Kazam_screencast_00038.webm](https://github.com/status-im/status-desktop/assets/20650004/d9bd33f8-ab43-43a8-8173-008ffb5524ff)

3. It's old issue but still alive - link elements affects formatting:

[Kazam_screencast_00037.webm](https://github.com/status-im/status-desktop/assets/20650004/74330fd8-0270-44dd-b21b-a507649b0630)

5. After pasting link after loading there is "Failed to generate preview" visible for some time:

[Kazam_screencast_00036.webm](https://github.com/status-im/status-desktop/assets/20650004/fb929b50-fac0-444c-8b78-38008b4e1619)

6. Ctrl+Z doesn't work after pasting link or uploading image

Other:
- it would be super nice to have backend mocked in storybook (providing some dummy images for links). The it could work as in the app, of course with dummy images. Using combobox for selecting number of images and links is a bit detached from the real usage.
- it would be nice to have the boxes scrollable with regular scroll, with no need to hold cmd button

alexjba commented on src/app/modules/shared_models/link_preview_model.nim:
```diff
@@ -105,33 +109,106 @@ QtObject:
+      return
+
+    self.items[index].linkPreview = initLinkPreview(link)
+    self.items[index].unfurled = false
```
hmm, it's not unfurled anymore. I mean yes, we have the link preview data in the cache and can be used somewhere else. But IMO when removing the link preview, the model item should look like no link preview has been ever loaded.

igor-sirotin replied:
ok, that makes sense 👍

alexjba commented on src/app/modules/shared_models/link_preview_item.nim:
```diff
@@ -4,6 +4,7 @@ import ../../../app_service/service/message/dto/link_preview
 type
   Item* = ref object
     unfurled*: bool
+    immutable*: bool
```
`immutable` describes the model item behaviour. `removed` describes the row state and the behaviour needs to be deduced by reading the implementation. Eg. It's not clear for a `removed` row if it can be added back, why it's still in the model or if it can be updated after being removed.

For `immutable` I find it clear by reading the public API that it's a row that can't be updated.

alexjba commented on src/app/modules/shared_models/link_preview_model.nim:
```diff
@@ -105,33 +109,106 @@ QtObject:
+    for url in urls:      
+      let linkPreview = initLinkPreview(url)
+      if(self.urlExists(url)):
+        itemsToUpdate[url] = linkPreview
```
Good point. Goes hand-in-hand with the comment below regarding the URLs order.
I've updated the implementation to use the following steps:
1. remove rows
2. Move or insert

There's no update at this step.

alexjba commented on ui/imports/shared/status/StatusChatInput.qml:
```diff
@@ -1306,6 +1314,8 @@ Rectangle {
                                 if(mention) {
                                     select(mention.leftIndex, mention.rightIndex)
+                                } else {
+                                    Global.openLink(link)
```
You're right. It shouldn't 👍

alexjba commented:
@micieslak I've updated the PR. Please have a look. Some issues are generic, reproducible on master and with tasks already logged. The CTRL+Z support was the trickiest one

> 1. When editing message and deleting link, it ends up with segfault:
> 
>  Kazam_screencast_00038.webm 

I've logged an issue on it. Looks like a status-go crash and it's reproducible on 0.14 also. https://github.com/status-im/status-desktop/issues/12323

> 3. It's old issue but still alive - link elements affects formatting:

There is one already for milestone 0.18: https://github.com/status-im/status-desktop/issues/9698
I think we need to revisit the syntax highlighter. Couldn't find a quick solution to this

> 5. After pasting link after loading there is "Failed to generate preview" visible for some time:

I can't reproduce. I've rebased on latest master and could include some status-go fixes as well.

> 6. Ctrl+Z doesn't work after pasting link or uploading image

Ohh, that's a tricky one! Took me a while to investigate the Qt support for undo stacks, but couldn't find any suitable option. I've added a custom undoStack to be used in tandem with the `TextEditHyperlinksFormatter`. Didn't see any other option because once the inner html text is altered, the Qt undo stack gets cleared. Please have a look on `UndoStackManager.qml`

> * it would be super nice to have backend mocked in storybook (providing some dummy images for links). The it could work as in the app, of course with dummy images. Using combobox for selecting number of images and links is a bit detached from the real usage.

Done. I've added some basic https detection.

> * it would be nice to have the boxes scrollable with regular scroll, with no need to hold cmd button

WDYM? On my end the scroll + drag support for the link preview cards is working fine.

alexjba commented on ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml:
```diff
@@ -0,0 +1,153 @@
+                                    unfurled && hostname === "" ? LinkPreviewMiniCard.State.LoadingFailed :
+                                    !unfurled ? LinkPreviewMiniCard.State.Loading : LinkPreviewMiniCard.State.Invalid
+
+                        onClose: root.linkRemoved(url)
```
Good point! I've update to use index + the remove action is done internally in the model. 

There's one other update related to this where I'd appreciate your feedback:
When sending the message, the link previews are fetched from the view model and included in the `sendChatMessage`. This is needed to exclude discarded link previews from the message

alexjba commented:
All the changes resulted from the review are included in the last commit: https://github.com/status-im/status-desktop/pull/12262/commits/4eaecfd88f9fa98632d29175ea2f8bea496dd85c

alexjba commented:
> > * it would be nice to have the boxes scrollable with regular scroll, with no need to hold cmd button
> 
> WDYM? On my end the scroll + drag support for the link preview cards is working fine.

@micieslak I see now. I was scrolling using the touchpad. The mouse scroll is not working. On it!

micieslak commented:
It looks good to me, but I found one small regression in comparison to master.

Undo on master:

[Kazam_screencast_00047.webm](https://github.com/status-im/status-desktop/assets/20650004/11dd6103-174e-4a2a-9f60-34ed6837ce8b)

When CTRL+Z is pressed, the addition in the middle is removed (the whole block abc, what seems to be a desired behavior) and cursor remains in the middle - also ok.

Undo in the PR:

[Kazam_screencast_00048.webm](https://github.com/status-im/status-desktop/assets/20650004/6227450f-67f8-4a3a-b2d8-ae3519464f63)

When CTRL+Z is pressed, only last letter from the addition in the middle is removed. The cursor always jumps to the end. Those two differences in comparison to the master seems to be undesired.

Regarding that I wonder if the custom undo stack makes sense - it fixes some rare cases (adding links) but breaks commonly expected behavior in general.

micieslak reviewed (changes requested):
One more think I've just noticed:

[Kazam_screencast_00049.webm](https://github.com/status-im/status-desktop/assets/20650004/ab136cda-e7bc-4703-b70c-bddf5fec6dcb)

Adding links seem to make behavior on switching between conversation inconsistent. Depending if we have links the content is preserved in conversation or replaced. It may easily lead to unintentional lost of message.

alexjba commented:
> One more think I've just noticed:
> 
>  Kazam_screencast_00049.webm 
> 
> Adding links seem to make behavior on switching between conversation inconsistent. Depending if we have links the content is preserved in conversation or replaced. It may easily lead to unintentional lost of message.

I have another pending  PR that sits on top of this where this is fixed. I've thought the changes made there introduced this bug. https://github.com/status-im/status-desktop/pull/12380

micieslak commented:
> > One more think I've just noticed:
> > Kazam_screencast_00049.webm
> > Adding links seem to make behavior on switching between conversation inconsistent. Depending if we have links the content is preserved in conversation or replaced. It may easily lead to unintentional lost of message.
> 
> I have another pending PR that sits on top of this where this is fixed. I've thought the changes made there introduced this bug. #12380

Cool, it works nice in https://github.com/status-im/status-desktop/pull/12380 indeed. So, no more questions, approving :)

micieslak reviewed (approved):
Great job!

caybro commented on ui/imports/shared/controls/chat/CalloutCard.qml:
```diff
@@ -1,34 +1,34 @@
+        path.fillColor: root.backgroundColor
+        path.strokeColor: root.borderColor
+        path.strokeWidth: root.borderWidth
+        path.strokeStyle: root.dashedBorder ? ShapePath.DashLine : ShapePath.SolidLine
```
Hmm, I already fixed that one :)

![image](https://github.com/status-im/status-desktop/assets/5377645/d3abfa95-6dcd-493d-bcd6-ea51608d4bbe)

caybro commented on ui/imports/utils/UndoStackManager.qml:
```diff
@@ -0,0 +1,158 @@
+            const newStackSize = Math.ceil(root.maxStackSize / 2)
+            print("Reducing undo stack to " + newStackSize + " items")
+            for(var i = 1; i <= newStackSize; i++) {
+                print("Removing " + Math.ceil(root.maxStackSize / newStackSize) + " items from index " + i)
```
Debug leftover?

caybro commented on ui/imports/shared/status/StatusChatInputImageArea.qml:
```diff
@@ -4,27 +4,31 @@ import QtQuick.Controls 2.13
     Repeater {
         id: rptImages
-
+    
```
Empty stray space?

caybro commented on ui/imports/shared/controls/chat/LinkPreviewMiniCard.qml:
```diff
@@ -0,0 +1,249 @@
+                        icon: "tiny/chevron-right"
+                        color: Theme.palette.baseColor1
+                        visible: secondTitleText.visible
+
```
empty line?

caybro commented on ui/app/AppLayouts/Chat/views/ChatColumnView.qml:
```diff
@@ -328,7 +304,7 @@ Item {
                              Global.playSendMessageSound()
 
-                             chatInput.textInput.clear();
+                             chatInput.setText("")
```
What's wrong with `clear()`?

caybro commented on ui/app/AppLayouts/Chat/views/ChatColumnView.qml:
```diff
@@ -133,15 +133,14 @@ Item {
 
             if (!d.activeChatContentModule) {
-                chatInput.textInput.text = ""
+                chatInput.setText("")
```
`clear()`? :)

caybro reviewed (approved):
Great job!

Some minor stuff inline, feel free to fixup in the followup PR
